### PR TITLE
Feat(ingest/ldap)fix list index out of range error

### PR DESCRIPTION
**Pull request by alplatonov**
issue  https://github.com/datahub-project/datahub/issues/8523

## Checklist
- [+] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [+] Links to related issues (if applicable)
- [+] Tests for the changes have been added/updated (if applicable)
- [+] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [+] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

Provided changes:
– Some fields may not be filled in, now there will be a check via `dict.get("key")`
– In some cases, the LDAP provider does not have pagination functionality, and the ingest fails, although all data was loaded successfully. Now possible to use the `pagination_enabled` option to work with LDAP providers.
– In LDAP ingestor, the `manager_pagination_enabled` changed to general `pagination_enabled`. The old option `manager_pagination_enabled` now provides a warning massage.

### CODE REVIEW

File: metadata-ingestion/tests/integration/ldap/ldap_mces_golden.json [hunk 1]
```diff
@@ -55,41 +55,6 @@
         "runId": "ldap-test"
     }
 },
-{
```
Comment (mayurinehate) [inline metadata-ingestion/tests/integration/ldap/ldap_mces_golden.json:58]: Why this got removed ?

Comment (alplatonov) [reply metadata-ingestion/tests/integration/ldap/ldap_mces_golden.json]: This is an error. I fix it.

File: metadata-ingestion/src/datahub/ingestion/source/ldap.py [hunk 2]
```diff
@@ -135,13 +137,41 @@ class LDAPSourceConfig(StatefulIngestionConfigBase, DatasetSourceConfigMixin):
 
     manager_pagination_enabled: bool = Field(
         default=True,
-        description="Use pagination while search for managers (enabled by default).",
+        description="[deprecated] Use pagination_enabled ",
+    )
+    _deprecate_manager_pagination_enabled = pydantic_field_deprecated(
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/ldap.py:142]: Why is this required ? There is also manager_filter_enabled. Does it also need renaming ?

Suggestion : Use pydantic_renamed_field instead. No need to write your own validator in that case.

File: metadata-ingestion/src/datahub/ingestion/source/ldap.py [hunk 3]
```diff
@@ -278,15 +311,17 @@ def get_workunits_internal(self) -> Iterable[MetadataWorkUnit]:
                 else:
                     self.report.report_dropped(dn)
 
-            pctrls = get_pctrls(serverctrls)
-            if not pctrls:
-                self.report.report_failure(
-                    "ldap-control", "Server ignores RFC 2696 control."
-                )
+            if serverctrls:
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/ldap.py:314]: Should the source report failure if serverctrls is not truthy ? When is serverctls not truthy ?

File: metadata-ingestion/src/datahub/ingestion/source/ldap.py [hunk 4]
```diff
@@ -300,15 +335,11 @@ def handle_user(self, dn: str, attrs: Dict[str, Any]) -> Iterable[MetadataWorkUn
...
-                if self.config.manager_pagination_enabled:
-                    ctrls = [self.lc]
-                else:
-                    ctrls = None
                 manager_msgid = self.ldap_client.search_ext(
                     m_cn,
                     ldap.SCOPE_BASE,
                     manager_filter,
-                    serverctrls=ctrls,
+                    serverctrls=[self.lc],
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/ldap.py:342]:
```suggestion
                    serverctrls=[self.lc] if self.lc else [],
```

File: metadata-ingestion/src/datahub/ingestion/source/ldap.py [hunk 5]
```diff
@@ -351,36 +382,42 @@ def build_corp_user_mce(
...
-        title = (
-            attrs[self.config.user_attrs_map["title"]][0].decode()
-            if self.config.user_attrs_map["title"] in attrs
-            else None
-        )
+        # We check two things:
+        # - attribute exists
+        # - attribute is not empty list
+        if attrs.get(self.config.user_attrs_map["email"]):
+            email = (attrs[self.config.user_attrs_map["email"]][0]).decode()
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/ldap.py:389]: can this logic be extracted into a method and use the method - say - `get_attr_or_none` everywhere. looks like a lot of repetition here..

```
def get_attr_or_none(attrs, key, default=None):
    return attrs[self.config.user_attrs_map[key]][0].decode() if attrs.get(self.config.user_attrs_map[key]) else default
```

also is it assumed that `attrs[self.config.user_attrs_map[key]]` is instance of  list ? Is is always true ? May help to add a check here to avoid errors.

File: metadata-ingestion/src/datahub/ingestion/source/ldap.py [hunk 6]
```diff
@@ -420,21 +457,24 @@ def build_corp_group_mce(self, attrs: dict) -> Optional[MetadataChangeEvent]:
...
-                if self.config.group_attrs_map["description"] in attrs
-                else None
-            )
-            displayName = (
-                attrs[self.config.group_attrs_map["displayName"]][0].decode()
-                if self.config.group_attrs_map["displayName"] in attrs
-                else None
-            )
+
+            if attrs.get(self.config.group_attrs_map["email"]):
```
Comment (mayurinehate) [inline metadata-ingestion/src/datahub/ingestion/source/ldap.py:461]: same comment as above

**Review by mayurinehate — CHANGES_REQUESTED** [review]

File: metadata-ingestion/src/datahub/ingestion/source/ldap.py (hunk 2 above)
Comment (alplatonov) [reply metadata-ingestion/src/datahub/ingestion/source/ldap.py:142]: > Why is this required ?

LDAP service like database, we can get error "SIZELIMIT_EXCEEDED: {'desc': 'Size limit exceeded'}"
This error tells us that we reached the limit of search result entries (1000 by default). There are 2 ways of resolving this error: 
- Go to your LDAP DATABASE and increase the maximum page size (MaxPageSize) option. 
- Use result paginating in your Python code.

Also, LDAP DATABASE sometimes doesn't support pagination at all. For example, google workspace LDAP doesn't support it -> so we need the option to turn off pagination completely, or we get an error.

Sometimes finding a person manager provides unexpected data and can be reduced with pagination.

> There is also manager_filter_enabled. Does it also need renaming?

No, this option is responsible for additional manager discovery. We want that person's head will be founded or not.

> Suggestion : Use pydantic_renamed_field instead. No need to write your own validator in that case.

Thx, I change it.

Comment (mayurinehate) [reply metadata-ingestion/src/datahub/ingestion/source/ldap.py:142]: Right. My question was around - why this rename is required - manager_pagination_enabled->pagination_enabled. Does this rename also warrant a rename in manager_filter_enabled->filter_enabled ?

File: metadata-ingestion/src/datahub/ingestion/source/ldap.py (hunk 3 above)
Comment (alplatonov) [reply metadata-ingestion/src/datahub/ingestion/source/ldap.py:314]: > Should the source report failure if `serverctrls` is not truth ?
No, only if enabled `pagecontrol`. i fix it.

> When is serverctls not truth?
In some scenarios, clients might want to check if the server supports specific controls without actually using them. To do this, they can send an empty server control in the request and observe the server's response. If the server supports the control, it will be included in the response.

Changed on `self.lc`

File: metadata-ingestion/src/datahub/ingestion/source/ldap.py (hunk 2 above)
Comment (alplatonov) [reply metadata-ingestion/src/datahub/ingestion/source/ldap.py]: In the old code/way, we always setup page control with line `self.lc = create_controls(self.config.page_size)`.
Sometimes LDAP provider doesn't support pagination. That's why i need an option – enable pagination or not in new code.

Now I have a decision:
1) create a new variable, for example `pagination_enabled` and `save manager_pagination_enabled`
2) rename `manager_pagination_enabled` to `pagination_enabled`

I choose option 2) because I don't see a situation where we need pagination for manager discovery, and at the same time, we need a global user/group pagination. If we have a problem and pagination is needed for the user/group or head, we can enable or disable it globally. The user will need clarification, as I am, with the name manager_pagination_enabled and global enable/disable pagination.

**Review by mayurinehate — APPROVED** [review]
lgtm

**Review by asikowitz — APPROVED** [review]
